File: scripts/release_date.pl
#!/usr/bin/env perl

use lib '../omp-perl';

use DateTime;
use OMP::DateTools;

use strict;

local $\ = "\n";
my $year_plus_day = DateTime::Duration->new(
    years => 1, hours => 23, minutes => 59, seconds => 59);

while (<>) {
    chomp;

    # Code modified from JSA::EnterData.

    my $semester = OMP::DateTools->determine_semester(
        date => $_, tel => 'UKIRT');
    my ($sem_begin, $sem_end) = OMP::DateTools->semester_boundary(
        semester => $semester, tel => 'UKIRT');

    my $release_date = DateTime->from_epoch(
        epoch => $sem_end->epoch, time_zone => 'UTC') + $year_plus_day;

    print $release_date->strftime('%Y%m%d %H:%M:%S');
    flush STDOUT;
}

File: lib/ukirt2caom2/release_date.py
from datetime import datetime

from taco import Taco

class ReleaseCalculator():
    def __init__(self):
        self.taco = Taco(lang='perl')
        self.taco.import_module('lib', '../omp-perl')
        self.taco.import_module('OMP::DateTools')

    def calculate(self, date):
        semester = self.taco.call_function(
            'OMP::DateTools::determine_semester', None,
            date=date.strftime('%Y%m%d'), tel='UKIRT')
        (sem_begin, sem_end) = self.taco.call_function(
            'OMP::DateTools::semester_boundary', None,
            semester=semester, tel='UKIRT', context='list')

        end_date = datetime.utcfromtimestamp(sem_end.call_method('epoch'))

        return end_date.replace(year=end_date.year + 1,
                                hour=23, minute=59, second=59)
